src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,414 @@
+
+            var (effects, temps) = initialEffectsAndTemps.ToImmutableAndFree();
+            var result = _factory.Sequence(temps, effects, returnValue);
+            return result;
+        }
+
+        private BoundExpression RewriteTupleOperator(TupleBinaryOperatorInfo @operator,
+            BoundExpression left, BoundExpression right, TypeSymbol boolType,
+            TupleOperatorSideEffectsAndTemps initialEffectsAndTemps, BinaryOperatorKind operatorKind, bool inLeftLiteral, bool inRightLiteral)
+        {
+            if (@operator.IsSingle())
+            {
+                return RewriteTupleSingleOperator((TupleBinaryOperatorInfo.Single)@operator, left, right, boolType, operatorKind);
+            }
+            else
+            {
+                return RewriteTupleNestedOperators((TupleBinaryOperatorInfo.Multiple)@operator, left, right, boolType, initialEffectsAndTemps, operatorKind, inLeftLiteral, inRightLiteral);
+            }
+        }
+
+        private BoundExpression RewriteTupleNestedOperators(TupleBinaryOperatorInfo.Multiple operators, BoundExpression left, BoundExpression right,
+            TypeSymbol boolType, TupleOperatorSideEffectsAndTemps initialEffectsAndTemps, BinaryOperatorKind operatorKind, bool inLeftLiteral, bool inRightLiteral)
+        {
+            // If either left or right is nullable, produce:
+            //
+            //      // outer sequence
+            //      var leftHasValue = left.HasValue; (or true if !leftNullable)
+            //      var rightHasValue = right.HasValue; (or true if !rightNullable)
+            //      leftHasValue == rightHasValue
+            //          ? leftHasValue ? ... inner sequence ... : true/false
+            //          : false/true
+            //
+            // where inner sequence is:
+            //      var leftValue = left.GetValueOrDefault(); (or left if !leftNullable)
+            //      var rightValue = right.GetValueOrDefault(); (or right if !rightNullable)
+            //      ... logical expression using leftValue and rightValue ...
+            //
+            // and true/false and false/true depend on operatorKind (== vs. !=)
+            //
+            // and temps are created an initialized with early side-effects for:
+            //  - elements from left and right, in case of a tuple literal
+            //  - left and right, in case of a tuple or nullable tuple type
+            //
+            // But if neither is nullable, then just produce the inner sequence.
+            //
+            // Note: all the temps are created in a single bucket (rather than different scopes of applicability) for simplicity
+
+            var outerEffects = ArrayBuilder<BoundExpression>.GetInstance();
+            var innerEffects = ArrayBuilder<BoundExpression>.GetInstance();
+
+            BoundExpression leftHasValue;
+            BoundExpression leftValue;
+
+            var coreLeft = WithoutImplicitNullableConversions(left);
```

Comment:
>WithoutImplicitNullableConversions [](start = 27, length = 34)

It feels like we are reinventing a "wheel" here, we already have helpers NullableNeverHasValue/NullableAlwaysHasValue tested and proven to do the right thing

Reply:
I looked at those methods, and neither seems to fit the need here. I added a PROTOTYPE to unblock.

---
In reply to: [171074477](https://github.com/dotnet/roslyn/pull/23356#discussion_r171074477) [](ancestors = 171074477)